README.md
# mccache | github.com/abbeymart/mccache-rs

- Simple and Hash In-Memory Information Cache
- Simple Cache: key-value pair storage
- Hash Cache: hash-hash-value storage
- See the test folder for test cases / scenarios and usage

## API - Simple Cache

- Simple Cache: key-value caching for rust-lang supported generic types - e.g. number, string and object values
- It exposes the following functions: set_cache, get_cache, delete_cache and clear_cache

## API - Hash Cache

- Hash Cache: hash-key-value storage for caching for rust-lang supported generic types - e.g. number, string and object values
- It exposes the following functions: set_hash_cache, get_hash_cache, delete_hash_cache and clear_hash_cache

## Usage

```rust

use crate::cache_types::*;

// simple-cache
use crate::simple_cache::*;

// hash-cache
use crate::hash_cache::*;


// See the test folder for different test cases / scenarios and usage
```


HISTORY.md
# Historical logs

## Version 0.1.0

- Release date: July-30-2023
- Simple and Hash In-Memory Cache
- See the test folder for different test cases / scenarios and usage


src/simple_cache.rs
use std::sync::Mutex;
use crate::cache_types::*;
use std::time::{SystemTime, UNIX_EPOCH};

// Initialise shared simple-cache object/dictionary (map)
pub static mut MC_SIMPLE_CACHE: Mutex<SimpleCache<T>> = Mutex::new(SimpleCache::new());

pub unsafe fn set_cache<T>(key: String, value: T, expire: u128) -> CacheResponseType<T> {
    // validate required params
    if key == "" || value == None {
        return CacheResponseType {
            ok: false,
            message: "cache-key and value are required".to_string(),
            value: (),
        };
    }
    // expire default Value (in seconds)
    let mut cache_expire = expire.clone();
    if expire == 0 {
        cache_expire = 300;
    }

    // current time in millisecond to set cache expiry time
    let mut current_time_in_milli_seconds: u128 = 0;
    match SystemTime::now().duration_since(UNIX_EPOCH) {
        Some(val) => {
            current_time_in_milli_seconds = val.as_millis();
        }
        _ => {
            let msg: String = format!("unable to set cache value - error setting current time for cache-expiration");
            CacheResponseType { ok: true, message: msg, value: () }
        }
    }
    // TODO: review unsafe variable | e.g. use storage singleton
    // compute secure cache-key
    let cache_key = format!("{}{}", key, KEY_CODE);
    // let cache_key_clone = cache_key.clone();
    // TODO: transform expire u64 to expire + time.Now
    let cache_value = CacheValueType { value, expire: cache_expire * 1000 + current_time_in_milli_seconds };
    // MC_SIMPLE_CACHE.lock().unwrap().insert(cache_key_clone, cache_value);

    match MC_SIMPLE_CACHE.lock() {
        Ok(mut cVal) => {
            // *cVal[cache_key] = cache_value.clone();
            cVal.insert(cache_key, cache_value.clone())
        }
        _ => {
            let msg: String = format!("unable to obtain cache lock");
            CacheResponseType { ok: true, message: msg, value: () }
        }
    }
    // validate value caching
    match MC_SIMPLE_CACHE.get(&cache_key) {
        Some(val) => {
            let msg: String = format!("task completed successfully");
            CacheResponseType { ok: true, message: msg, value: val }
        }
        None => {
            let msg: String = format!("unable to set cache value");
            CacheResponseType { ok: true, message: msg, value: () }
        }
    }
}

pub unsafe fn get_cache<T>(key: String) -> CacheResponseType<T> {
    // validate required params
    if key == "" {
        return CacheResponseType {
            ok: false,
            message: "cache-key is required".to_string(),
            value: (),
        };
    }
    // compute secure cache-key
    let cache_key = format!("{}{}", key, KEY_CODE);
    // current time in millisecond to set cache expiry time
    let mut current_time_in_milli_seconds: u128 = 0;
    match SystemTime::now().duration_since(UNIX_EPOCH) {
        Some(val) => {
            current_time_in_milli_seconds = val.as_millis();
        }
        _ => {
            let msg: String = format!("unable to set cache value - error setting current time for cache-expiration");
            CacheResponseType { ok: true, message: msg, value: () }
        }
    }
    // fetch cache value from shared memory storage
    match MC_SIMPLE_CACHE.get(&cache_key) {
        Some(val) => if val.value != None && val.expire > current_time_in_milli_seconds {
            let msg: String = format!("task completed successfully");
            CacheResponseType { ok: true, message: msg, value: val }
        } else if val.value != None && val.expire < current_time_in_milli_seconds {
            match MC_SIMPLE_CACHE.lock() {
                Ok(mut cVal) => {
                    cVal.remove(&cache_key);
                    let msg: String = format!("cache expired and deleted");
                    CacheResponseType { ok: true, message: msg, value: () }
                }
                _ => {
                    let msg: String = format!("unable to obtain cache lock");
                    CacheResponseType { ok: true, message: msg, value: () }
                }
            }
        } else {
            let msg: String = format!("cache info does not exist");
            CacheResponseType { ok: true, message: msg, value: () }
        },
        None => {
            let msg: String = format!("cache info does not exist");
            CacheResponseType { ok: true, message: msg, value: () }
        }
    }
}

pub unsafe fn delete_cache<T>(key: String) -> CacheResponseType<T> {
    // validate required params
    if key == "" {
        return CacheResponseType {
            ok: false,
            message: "cache-key is required".to_string(),
            value: (),
        };
    }
    let cache_key = format!("{}{}", key, KEY_CODE);

    match MC_SIMPLE_CACHE.lock() {
        Ok(mut cVal) => {
            // *cVal[cache_key] = None;
            match cVal.remove(&cache_key) {
                Some(val) => {
                    let msg: String = format!("task completed successfully");
                    CacheResponseType { ok: true, message: msg, value: val }
                }
                None => {
                    let msg: String = format!("task not completed, cache-key-value not found");
                    CacheResponseType { ok: true, message: msg, value: () }
                }
            }
        }
        _ => {
            let msg: String = format!("unable to obtain cache lock");
            CacheResponseType { ok: true, message: msg, value: () }
        }
    }


}

pub unsafe fn clear_cache<T>() -> CacheResponseType<T> {
    match MC_SIMPLE_CACHE.lock() {
        Ok(mut cVal) => {
            // *cVal = SimpleCache::new();
            cVal.clear();
            let msg: String = format!("task completed successfully");
            CacheResponseType { ok: true, message: msg, value: () }
        }
        _ => {
            let msg: String = format!("unable to obtain cache lock");
            CacheResponseType { ok: true, message: msg, value: () }
        }
    }
}



src/cache_storage.rs
use crate::cache_types::*;

//secret keyCode for added security
pub const KEY_CODE: &str = "mcconnect_20200320";

// TODO: review variable generic <T> setting and use of unsafe static mut variables | consider safe alternative

// Initialise simple-cache object/dictionary (map)
pub static mut MC_SIMPLE_CACHE: SimpleCache<T> = SimpleCache::new();

// Initialise hash-cache object/dictionary (map)
pub static mut MC_HASH_CACHE: HashCache<T> = HashCache::new();

// pub unsafe fn cache_storage<T>(mc_cache: &mut SimpleCache<T>) -> &SimpleCache<T> {
// Create singleton cache storage for simple-cache - at application level
pub unsafe fn cache_storage<T>() -> &SimpleCache<T> {
    let mut mc_simple_cache = SimpleCache::new();
    return &mc_simple_cache
}

// Create singleton cache storage for hash-cache - at application level
pub fn hash_cache_storage<T>() -> &HashCache<T> {
    let mut mc_hash_cache = HashCache::new();
    return &mc_hash_cache
}

src/hash_cache.rs
use std::sync::Mutex;
use crate::cache_types::*;
use std::time::{SystemTime, UNIX_EPOCH};

// Initialise shared hash-cache object/dictionary (map)
pub static mut MC_HASH_CACHE: Mutex<HashCache<T>> = Mutex::new(HashCache::new());

pub unsafe fn set_hash_cache<T>(key: String, hash: String, value: T, expire: u128) -> CacheResponseType<T> {
    // validate required params
    if key == "" || hash == "" || value == None {
        return CacheResponseType {
            ok: false,
            message: "hash, cache-key and value are required".to_string(),
            value: (),
        };
    }
    // expire default Value (in seconds)
    let mut cache_expire = expire.clone();
    if expire == 0 {
        cache_expire = 300;
    }
    // current time in millisecond to set cache expiry time
    let mut current_time_in_milli_seconds: u128 = 0;
    match SystemTime::now().duration_since(UNIX_EPOCH) {
        Some(val) => {
            current_time_in_milli_seconds = val.as_millis();
        }
        _ => {
            let msg: String = format!("unable to set cache value - error setting current time for cache-expiration");
            CacheResponseType { ok: true, message: msg, value: () }
        }
    }
    // TODO: review unsafe variable | e.g. use storage singleton
    // compute secure cache-key and hash-key
    let hash_key = format!("{}{}", hash, KEY_CODE);
    let cache_key = format!("{}{}", key, KEY_CODE);

    let mut hash_cache_value: HashCacheValueType<T> = HashCacheValueType::new();
    let cache_value = CacheValueType { value, expire: cache_expire * 1000 + current_time_in_milli_seconds };
    hash_cache_value.insert(cache_key, cache_value);
    MC_HASH_CACHE.insert(hash_key_clone, hash_cache_value);

    match MC_HASH_CACHE.lock() {
        Ok(mut cVal) => {
            // *cVal[hash_key] = hash_cache_value.clone();
            cVal.insert(hash_key, hash_cache_value.clone());
            // MC_HASH_CACHE.insert(hash_key, hash_cache_value.clone());
        }
        _ => {
            let msg: String = format!("unable to obtain cache lock");
            CacheResponseType { ok: true, message: msg, value: () }
        }
    }

    // validate value caching
    match MC_HASH_CACHE.get(&cache_key) {
        Some(val) => {
            let msg: String = format!("task completed successfully");
            CacheResponseType { ok: true, message: msg, value: val }
        }
        Err(_) => {
            let msg: String = format!("unable to set cache value");
            CacheResponseType { ok: true, message: msg, value: val }
        }
    }
}

pub unsafe fn get_hash_cache<T>(key: String, hash: String) -> CacheResponseType<T> {
    // validate required params
    if key == "" || hash == "" {
        return CacheResponseType {
            ok: false,
            message: "hash and cache-key are required".to_string(),
            value: (),
        };
    }
    // compute secure cache-key
    let hash_key = format!("{}{}", hash, KEY_CODE);
    let cache_key = format!("{}{}", key, KEY_CODE);
    // current time in millisecond to set cache expiry time
    let mut current_time_in_milli_seconds: u128 = 0;
    match SystemTime::now().duration_since(UNIX_EPOCH) {
        Some(val) => {
            current_time_in_milli_seconds = val.as_millis();
        }
        _ => {
            let msg: String = format!("unable to set cache value - error setting current time for cache-expiration");
            CacheResponseType { ok: true, message: msg, value: () }
        }
    }

    match MC_HASH_CACHE.get(&hash_key) {
        Some(kVal) => match kVal.get(&cache_key) {
            Some(val) => if val.value != None && val.expire > current_time_in_milli_seconds {
                let msg: String = format!("task completed successfully");
                CacheResponseType { ok: true, message: msg, value: val }
            } else if val.value != None && val.expire < current_time_in_milli_seconds {
                // remove cache-item
                match MC_HASH_CACHE.lock() {
                    Ok(mut cVal) => {
                        // *cVal[hash_key] = hash_cache_value.clone();
                        match cVal.get(&hash_key).remove(&cache_key) {
                            Some(val) => {
                                let msg: String = format!("cache expired and deleted");
                                CacheResponseType { ok: true, message: msg, value: val }
                            }
                            None => {
                                let msg: String = format!("cache info does not exist");
                                CacheResponseType { ok: true, message: msg, value: () }
                            }
                        }
                    }
                    _ => {
                        let msg: String = format!("unable to obtain cache lock");
                        CacheResponseType { ok: true, message: msg, value: () }
                    }
                }
            } else {
                let msg: String = format!("cache info does not exist");
                CacheResponseType { ok: true, message: msg, value: () }
            },
            Err(_) => {
                let msg: String = format!("cache info does not exist");
                CacheResponseType { ok: true, message: msg, value: () }
            }
        },
        Err(_) => {
            let msg: String = format!("cache info does not exist");
            CacheResponseType { ok: true, message: msg, value: () }
        }
    }
}

pub unsafe fn delete_hash_cache<T>(key: String, hash: String, by: String) -> CacheResponseType<T> {
    // validate required params
    if key == "" || hash == "" {
        return CacheResponseType {
            ok: false,
            message: "hash and cache-key are required".to_string(),
            value: (),
        };
    }
    let mut del_by = "key".to_string();
    if by != "key" {
        del_by = by.to_string();
    }
    if key == "" || hash == "" && by == "key" {
        return CacheResponseType {
            ok: false,
            message: "hash and cache keys are required".to_string(),
            value: (),
        };
    }
    if hash == "" && by == "hash" {
        return CacheResponseType {
            ok: false,
            message: "hash key is required".to_string(),
            value: (),
        };
    }
    // compute secure cache-key
    let hash_key = format!("{}{}", hash, KEY_CODE);
    let cache_key = format!("{}{}", key, KEY_CODE);

    if del_by == "hash" {
        match MC_HASH_CACHE.lock() {
            Ok(mut cVal) => {
                // *cVal[hash_key] = None;
                match cVal.remove(&hash_key) {
                    Some(val) => {
                        let msg: String = format!("task completed successfully");
                        CacheResponseType { ok: true, message: msg, value: val }
                    }
                    None => {
                        let msg: String = format!("task not completed, hash-value not found");
                        CacheResponseType { ok: true, message: msg, value: val }
                    }
                }
            }
            _ => {
                let msg: String = format!("unable to obtain cache lock");
                CacheResponseType { ok: true, message: msg, value: () }
            }
        }
    }

    if del_by == "key" {
        match MC_HASH_CACHE.lock() {
            Ok(cVal) => {
                match cVal.get(&hash_key) {
                    Some(_) => {
                        // *cVal[hash_key][cache_key] = None;
                        match cVal[hash_key].remove(&cache_key) {
                            Some(val) => {
                                let msg: String = format!("task completed successfully");
                                CacheResponseType { ok: true, message: msg, value: val }
                            }
                            None => {
                                let msg: String = format!("task not completed, hash-cache-value not found");
                                CacheResponseType { ok: true, message: msg, value: () }
                            }
                        }
                    }
                    None => {
                        let msg: String = format!("task not completed, hash-value not found");
                        CacheResponseType { ok: true, message: msg, value: () }
                    }
                }
            }
            _ => {
                let msg: String = format!("unable to obtain cache lock");
                CacheResponseType { ok: true, message: msg, value: () }
            }
        }
    }

    let msg: String = format!("invalid-by-params[key or hash]");
    CacheResponseType { ok: true, message: msg, value: nil }
}

pub unsafe fn clear_hash_cache<T>() -> CacheResponseType<T> {
    match MC_HASH_CACHE.lock() {
        Ok(mut cVal) => {
            // *cVal = HashCache::new();
            cVal.clear();
            let msg: String = format!("task completed successfully");
            CacheResponseType { ok: true, message: msg, value: () }
        }
        _ => {
            let msg: String = format!("unable to obtain cache lock");
            CacheResponseType { ok: true, message: msg, value: () }
        }
    }
}
